cpp/src/parquet/page_index.cc

```diff
@@ -184,8 +185,241 @@ class OffsetIndexImpl : public OffsetIndex {
+  std::shared_ptr<::arrow::Buffer> offset_index_buffer_;
+};
+
+class PageIndexReaderImpl : public PageIndexReader {
+ public:
+  PageIndexReaderImpl(::arrow::io::RandomAccessFile* input,
+                      std::shared_ptr<FileMetaData> file_metadata,
+                      const ReaderProperties& properties,
+                      std::shared_ptr<InternalFileDecryptor> file_decryptor)
+      : input_(input),
+        file_metadata_(std::move(file_metadata)),
+        properties_(properties),
+        file_decryptor_(std::move(file_decryptor)) {}
+
+  ::arrow::Result<std::shared_ptr<RowGroupPageIndexReader>> RowGroup(int i) override {
+    if (i < 0 || i >= file_metadata_->num_row_groups()) {
+      return ::arrow::Status::IndexError("Invalid row group ordinal {}", i);
+    }
+    return std::make_shared<RowGroupPageIndexReaderImpl>(
+        input_, file_metadata_->RowGroup(i), properties_, i, file_decryptor_);
+  }
+
+  void WillNeed(const std::vector<int32_t>& row_group_indices, bool need_column_index,
+                bool need_offset_index) override {
+    std::vector<::arrow::io::ReadRange> read_ranges;
+    for (int32_t row_group_ordinal : row_group_indices) {
+      int64_t column_index_offset;
+      int64_t column_index_size;
+      int64_t offset_index_offset;
+      int64_t offset_index_size;
+      bool has_column_index;
+      bool has_offset_index;
+      PageIndexReader::DeterminePageIndexRangesInRowGroup(
+          *file_metadata_->RowGroup(row_group_ordinal), &column_index_offset,
+          &column_index_size, &offset_index_offset, &offset_index_size, &has_column_index,
+          &has_offset_index);
+      if (need_column_index && has_column_index) {
+        read_ranges.emplace_back(
+            ::arrow::io::ReadRange{column_index_offset, column_index_size});
+      }
+      if (need_offset_index && has_offset_index) {
+        read_ranges.emplace_back(
+            ::arrow::io::ReadRange{offset_index_offset, offset_index_size});
+      }
+    }
+    PARQUET_IGNORE_NOT_OK(input_->WillNeed(read_ranges));
+  }
+
+  void WillNotNeed(const std::vector<int32_t>& row_group_indices) override {
+    // No-op for now.
+  }
+
+ private:
+  /// The input stream that can perform random read.
+  ::arrow::io::RandomAccessFile* input_;
+
+  /// The file metadata to get row group metadata.
+  std::shared_ptr<FileMetaData> file_metadata_;
+
+  /// Reader properties used to deserialize thrift object.
+  const ReaderProperties& properties_;
+
+  /// File-level decrypter.
+  std::shared_ptr<InternalFileDecryptor> file_decryptor_;
+};
+
 }  // namespace
 
+void PageIndexReader::DeterminePageIndexRangesInRowGroup(
+    const RowGroupMetaData& row_group_metadata, int64_t* column_index_start,
+    int64_t* column_index_size, int64_t* offset_index_start, int64_t* offset_index_size,
+    bool* has_column_index, bool* has_offset_index) {
+  int64_t ci_start = std::numeric_limits<int64_t>::max();
```

Comment:
per comment above consider making a byterange struct here (@pitrou do you recall if something like this exists already in parquet or Arrow)?

Reply:
What about `::arrow::io::ReadRange`?

Reply:
I have consolidated the return value to a single struct `RowGroupIndexReadRange` which internally leverages `::arrow::io::ReadRange`. It looks better now.